src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractItemFileWatchDataSource.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks.Dataflow;
+
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    /// <summary>
+    ///     Base class for producing items so that they can be watched, and when changed, trigger re-evaluation.
+    /// </summary>
+    internal abstract class AbstractItemFileWatchDataSource : ChainedProjectValueDataSourceBase<FileWatchData>, IFileWatchDataSource
+    {
+        private readonly ConfiguredProject _project;
+        private readonly IProjectSubscriptionService _projectSubscriptionService;
+
+        protected AbstractItemFileWatchDataSource(ConfiguredProject project, IProjectSubscriptionService projectSubscriptionService)
+            : base(project.Services, synchronousDisposal:true, registerDataSource: false)
+        {
+            _project = project;
+            _projectSubscriptionService = projectSubscriptionService;
+        }
+
+        public abstract string ItemRuleName
+        {
+            get;
+        }
+
+        public abstract FileWatchChangeKinds FileWatchChangeKinds
+        {
+            get;
+        }
+
+        protected override UnconfiguredProject ContainingProject
+        {
+            get { return _project.UnconfiguredProject; }
+        }
+
+        protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<FileWatchData>> targetBlock)
+        {
+            IProjectValueDataSource<IProjectSubscriptionUpdate> source = _projectSubscriptionService.ProjectRuleSource;
+
+            // Transform the changes from evaluation/design-time build -> restore data
+            DisposableValue<ISourceBlock<IProjectVersionedValue<FileWatchData>>> transformBlock = source.SourceBlock
+                                                                                                        .TransformWithNoDelta(update => update.Derive(u => CreateFileWatch(u.CurrentState)),
+                                                                                                            suppressVersionOnlyUpdates: true,
+                                                                                                            ruleNames: new[] { ItemRuleName });
+
+            // Set the link up so that we publish changes to target block
+            transformBlock.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
```

Comment:
Do you need to dispose the link returned here? I've noticed there is a lot of inconsistencies around this in general.

Reply:
You need to dispose the first link, it's a bit confusing...